### ISSUE TITLE
question between qnn and snpe

### ISSUE THREAD

**Issue by cvv-student** [quic/qidk#47]
I have noticed that the engineering based on Android in this project is all accomplished using SNPE, while QNN is invoked through python. Could you elucidate the relationship between these qnn and snpe? Is it that the PC only utilizes QNN and Android utilizes SNPE, or can QNN also be employed on Android? If so, could provide a demo for invoking QNN on Android?

**Comment by quic-vraidu** [comment]
Please refer following links to understand the details: 
SNPE: https://docs.qualcomm.com/bundle/publicresource/topics/80-70014-15B/snpe.html
QNN: https://docs.qualcomm.com/bundle/publicresource/topics/80-70014-15B/qnn.html

QNN can run on Android. I dont have opensource reference example. We are planning to update QIDK github with QNN examples in coming weeks.

**Comment by cvv-student** [comment]
Could it be understood from the documents that QNN is oriented towards hardware development, offering low-level APIs for more versatile acceleration, whereas SNPE is model-oriented, relatively simplified and user-friendly in comparison to QNN? I hope to see an Android demo of QNN soon. I achieved decent speed using onnxruntime on Gen2 SOC (utilizing HTP), but the speed was rather slow when I individually called naive API of QNN.

**Comment by quic-vraidu** [comment]
Yes. Your understanding is correct. 
Are your model generated with HTP QNN?

**Comment by cvv-student** [comment]
Indeed, it was generated using the executable file of QNN. Upon comparing the performance differences between SNPE and QNN, I found them to be quite similar. Thus, I have decided to proceed with SNPE. The current issue is that the presence of these libraries of SNPE makes the APK substantially large. I intend to load them in a manner akin to caching. Specifically, after installing the APK, the runtime libraries will be pushed to a cache directory. Subsequently, the library path can be obtained using File targetDir = context.getDir("libs", context.MODE_PRIVATE);. The libraries will then be copyed in the targetDir folder, and dynamically loaded using the System.load function. However, an error occurs when loading libSnpeHtpV69Skel.so. Here is the log: Error message: dlopen failed: "/data/data/com.ai.test/app_libs/libSnpeHtpV69Skel.so" is 32-bit instead of 64-bit. I checked the file libSnpeHtpV69Skel.so and confirmed that it is indeed a 32-bit library. However, when I place this library in the jnilibs folder and install it along with the APK, it runs without issues. Naturally, I also encounter an error when attempting to load SnpeHtpV69Skel using System.loadLibrary, as it cannot load a 32-bit library. So, how should I dynamically load libSnpeHtpV69Skel.so?

**Comment by quic-vraidu** [comment]
libSnpeHtpV69Skel.so is only 32bit supported. Libraries. 
In stackoverflow, I can see few options, you can try it out. 
https://stackoverflow.com/questions/27186243/use-32-bit-jni-libraries-on-64-bit-android/27202958

**Comment by cvv-student** [comment]
Thank you for your reply. I attempted the aforementioned solution, but it still didn't truly resolve the issue. The file `libSnpeHtpV69Skel.so` is dynamically loaded by the SNPE runtime. When it is dynamically loaded as an external file, the log indicates that it is found, yet it fails to load properly. Here is the erroneous log:
```
vendor/qcom/proprietary/adsprpc/src/apps_std_imp.c:880: Successfully opened file libSnpeHtpV69Skel.so
vendor/qcom/proprietary/adsprpc/src/fastrpc_apps_user.c:1437: Error 0x80000442: remote_handle_open_domain: dynamic loading failed for file:///libSnpeHtpV69Skel.so?snpe_skel_handle_invoke&_modver=1.0&_dom=cdsp on domain 3 (dlerror RX VA 0xFF300000 outside ELF segment) (errno Success)           
vendor/qcom/proprietary/adsprpc/src/fastrpc_apps_user.c:1496: Error 0x80000442: remote_handle64_open failed for file:///libSnpeHtpV69Skel.so?snpe_skel_handle_invoke&_modver=1.0&_dom=cdsp (errno Success)
```

**Comment by quic-vraidu** [comment]
Please try to check below command enabled.
adb shell setenforce 0

**Comment by cvv-student** [comment]
Thank you. The execution of the aforementioned command was denied. Even if it can be executed locally, there is no guarantee that the actual user environment will have root privileges.

**Comment by quic-vraidu** [comment]
You can use below commands for root access.
adb root
adb remount

**Comment by cvv-student** [comment]
The command was rejected as lack root privileges. My goal is to run it on a machine without root access.

**Comment by quic-vraidu** [comment]
If Unsigned or Signed DSP runtime is not getting detected, then please check the logcat logs for the FastRPC error. DSP runtime may not get detected due to SE Linux security policy. Please try out following commands to set permissive SE Linux policy.

It is recommended to run below commands.

adb disable-verity
adb reboot
adb root
adb remount
adb shell setenforce 0
